geode-core/src/distributedTest/java/org/apache/geode/management/JMXMBeanLocatorRestartTest.java

```diff
@@ -0,0 +1,317 @@
+      new AtomicReference<>(new CountDownLatch(0));
+  private static final AtomicReference<CountDownLatch> AFTER =
+      new AtomicReference<>(new CountDownLatch(0));
+
+  private static final AtomicReference<LocatorLauncher> LOCATOR =
+      new AtomicReference<>(DUMMY_LOCATOR);
+  private static final AtomicReference<ServerLauncher> SERVER =
+      new AtomicReference<>(DUMMY_SERVER);
+
+  private VM locatorVM;
+  private VM serverVM;
+
+  private String locatorName;
+  private String serverName;
+  private String locators;
+  private int locatorPort;
+  private int locatorJmxPort;
+  private String regionName;
+  private Set<ObjectName> mxbeansOnServer;
+  private Set<ObjectName> mxbeansOnLocator;
+
+  @Rule
+  public DistributedRule distributedRule = new DistributedRule();
+  @Rule
+  public SerializableTemporaryFolder temporaryFolder = new SerializableTemporaryFolder();
+  @Rule
+  public DistributedErrorCollector errorCollector = new DistributedErrorCollector();
+  @Rule
+  public transient GfshCommandRule gfsh = new GfshCommandRule();
+  @Rule
+  public DistributedRestoreSystemProperties restoreProps = new DistributedRestoreSystemProperties();
+
+  @Before
+  public void setUp() throws Exception {
+    locatorVM = getVM(1);
+    serverVM = getVM(0);
+
+    for (VM vm : asList(locatorVM, serverVM)) {
+      vm.invoke(() -> System.setProperty(GEMFIRE_PREFIX + "standard-output-always-on", "true"));
+    }
+
+    locatorName = "locator1";
+    serverName = "server1";
+    File locator1Dir = temporaryFolder.newFolder(locatorName);
+    File serverDir = temporaryFolder.newFolder(serverName);
+
+    int[] port = getRandomAvailableTCPPorts(4);
+    locatorPort = port[0];
+    locatorJmxPort = port[2];
+    locators = "localhost[" + locatorPort + "]";
+
+    locatorVM.invoke(() -> {
+      startLocator(locatorName, locator1Dir, locatorPort, locatorJmxPort, locators);
+    });
+
+    serverVM.invoke(() -> startServer(serverName, serverDir, locators));
+
+    gfsh.connectAndVerify(locatorJmxPort, GfshCommandRule.PortType.jmxManager);
+
+    regionName = "region1";
+    String createRegionCommand = "create region --type=REPLICATE --name=" + SEPARATOR + regionName;
+    gfsh.executeAndAssertThat(createRegionCommand).statusIsSuccess();
+
+    addIgnoredException(AlertingIOException.class);
+    addIgnoredException(CacheClosedException.class);
+    addIgnoredException(CancelException.class);
+    addIgnoredException(DistributedSystemDisconnectedException.class);
+    addIgnoredException(ForcedDisconnectException.class);
+    addIgnoredException(MemberDisconnectedException.class);
+    addIgnoredException("Possible loss of quorum");
+
+    mxbeansOnServer = serverVM.invoke(() -> {
+      await().untilAsserted(() -> {
+        assertThat(getPlatformMBeanServer().queryNames(getInstance("GemFire:*"), null))
+            .as("GemFire mbeans on server1")
+            .containsAll(expectedServerMXBeans(serverName, regionName));
```

Comment:
Can the result of `expectedServerMXBeans(…)` change each time through the loop? If not, is it worth it to call it before the loop, rather than inside?

Similar for other inside-the-loop calls to get lists of expected beans.

Reply:
Yeah I should probably change those to be invoked within the await loop. They end up creating new instances of ObjectName every time they're called.

Reply:
I just remembered why I have it like this: for easier reading. I'm converting them all to variables and one field, but it's not as readable.

Reply:
Yeah, that's part of why I was wondering whether my suggestion was worth the effort.

Reply:
I think I found a good way to improve readability along with make things more efficient. Please take a look at both JmxLocatorReconnectDistributedTest (my new test) and JmxServerReconnectDistributedTest (overhauled for consistency). Thanks!

Reply:
Oops, I forgot to push earlier. The updated distributed tests are pushed now.